### PR TITLE
Fix pr3090 / Adds support for compute-resources

### PR DESCRIPTION
**Pull request by kbidarkar**
Pending: Need to add the configuration checks.

This is a very old PR and has code which was written before we re-factored a lot of stuff.
So the plan is to split this up into to 2 PR's to simplify this PR.
a) First CRUD PR: PR with REBASE, merge_conflicts, locators clean up, clean up of compute_resource CRUD functions and addressing the comments for the same.
b) Second RHCI tests: PR with rhci/tests, removal of ddt, data and implement subTest and address other comments for the same.

### CODE REVIEW

**Comment by kbidarkar** [comment]
So this PR includes, the below 
a) First CRUD PR: PR with REBASE, merge_conflicts, locators clean up, clean up of compute_resource CRUD functions, addition of settings checks for CR's  and addressing the comments for the same.

I tried to keep this PR as independent as possible, so that it gets merged easily :). The idea was this PR should probably not break any test but just get merged.

Also PR 3090 had tests for RHCI and it's perspective and had used @data and @ddt which is as good as re-writing the tests from scratch for normal CR related tests and also after all the re-factoring. But feel this being Comput-Resource Functional test-cases and quiet time consuming these should probably not be using subTests.

File: robottelo/ui/computeresource.py [hunk 1]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
+        self.click(locators['resource.virtual_machines_tab'])
+        vm_elements = self.find_element(locators['resource.vm_list'])
+        return [vm.text for vm in vm_elements]
+
+    def add_image(self, res_name, parameter_list):
+        """Adds an image to a compute resource."""
+        self.access_compute_resource(res_name)
+        self.click(locators['resource.image_add'])
+        self.wait_until_element(locators['resource.image_name'])
+        if parameter_list is None:
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: That can be a first line in the method

File: robottelo/ui/computeresource.py [hunk 2]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
+        self.click(locators['resource.image_submit'])
+        self.wait_until_element(common_locators['notif.success'])
+
+    def list_images(self, res_name):
+        """Lists images on Compute Resource.
+        Note: Currently lists only images that show up on the first page.
+        """
+        self.access_compute_resource(res_name)
+        self.click(locators['resource.images_tab'])
+        image_elements = self.find_element(locators['resource.image_list'])
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: Doesn't `self.find_element` always return only one/first element? You can look on the find_elements from Selenium library if you like to have list of WebElements

File: robottelo/ui/computeresource.py [hunk 3]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
+                'Could not find the resource {0}'.format(res_name))
+        self.click((strategy, value % res_name))
+
+    def list_vms(self, res_name):
+        """Lists vms of a particular compute resource.
+        Note: Currently lists only vms that show up on the first page.
+        """
+        self.access_compute_resource(res_name)
+        self.click(locators['resource.virtual_machines_tab'])
+        vm_elements = self.find_element(locators['resource.vm_list'])
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: Doesn't `self.find_element` always return only one/first element? You can look on the find_elements from Selenium library if you like to have list of WebElements

File: robottelo/ui/computeresource.py [hunk 4]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
-        """Removes the compute resource entity"""
+        """Removes the compute resource entity."""
         self.delete_entity(
             name,
             really,
             locators['resource.delete'],
             drop_locator=locators['resource.dropdown'],
         )
+
+    def access_compute_resource(self, res_name):
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: not sure why we need specific method for this operation. Doesn't `self.click(self.search(res_name))` serve the same purpose?

File: robottelo/ui/computeresource.py [hunk 5]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
+        button = self.find_element((strategy, value % (res_name, vm_name)))
+        if 'Off' not in button.text:
+            raise UIError(
+                'Could not stop VM {0}. VM is not running'.format(vm_name)
+            )
+        self.click(button)
+        self.handle_alert(really)
+        self.wait_until_element(common_locators['notif.success'])
+
+    def vm_action_start(self, res_name, vm_name):
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: Could you please look on similar method `set_power_status` from `container.py` for reference? Maybe we can use one method instead these three?

File: robottelo/ui/computeresource.py [hunk 6]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
+                (parameter_name.lower()).replace(' ', '_')
+            ))
+            if parameter_type == 'field':
+                self.find_element(
+                    locators[param_locator]
+                ).send_keys(parameter_value)
+            elif parameter_type == 'select':
+                self.select(locators[param_locator], parameter_value)
+        self.click(locators['resource.image_submit'])
+        self.wait_until_element(common_locators['notif.success'])
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: Usually we do not use such assertions on a framework level, but do them on a test level instead. That recommendation is applied for all methods in this module

**Comment by kbidarkar** [comment]
This PR is part of my/our WIG goals and would soon start working on it.

File: robottelo/ui/computeresource.py [hunk 7]
```diff
@@ -137,10 +135,107 @@ def update(self, name, newname=None, parameter_list=None,
...
         )
+
+    def access_compute_resource(self, res_name):
+        """Access a particular compute resource."""
+        strategy, value = locators['resource.get_by_name']
+        resource = self.search(res_name)
+        if resource is None:
+            raise UINoSuchElementError(
+                'Could not find the resource {0}'.format(res_name))
+        self.click((strategy, value % res_name))
```
Comment (rochacbruno) [inline robottelo/ui/computeresource.py]: Now you can use locators directly as: 
`self.click(locators['resource.get_by_name'] % res_name)`

Not a blocker, just a comment (in future we should change every occurrence)

Comment (kbidarkar) [inline robottelo/ui/computeresource.py]: Will pick this PR now and try to finish this hopefuly by monday, Thanks for the suggestions, I am away from the robottelo automation, I did see some changes from you around this but didn't get chance to go through in detail. So good to know the current usage.

File: robottelo/ui/computeresource.py (hunk 5 above)
Comment (kbidarkar) [inline robottelo/ui/computeresource.py]: Yeah, sure will explore that, but would plan to do that in a different PR, this PR is pending since a long time and want to get this merged ASAP now.

What I would do is , raise a new github issue and assign it to me and @san7ket . This PR once merged would also unblock @san7ket  from raising his PR's. 

Will fix that for sure as part of this, https://github.com/SatelliteQE/robottelo/issues/3754

**Comment by kbidarkar** [comment]
Sorry, missed fixing for flake8, Will be fixing travis now.

File: robottelo.properties.sample [hunk 8]
```diff
@@ -269,3 +270,57 @@ ssh_key=
 # for RHEL7 - the ports needs to have a correct selinux context set:
 # 'semanage port -a -t websm_port_t -p tcp <port-range>'
 port_range=9091, 14999
+
+# [rhev]
+# RHEV to be added as a compute resource.
+# hostname: RHEV API URL, for example: https://ovirt.example.com/api
```
Comment (elyezer) [inline robottelo.properties.sample]: What about placing the comment for each field on top of it?

Like:

```
# RHEV API URL, for example: https://ovirt.example.com/api
#hostname=

# Login for RHEVM
#username=

...
```

File: robottelo/ui/computeresource.py [hunk 9]
```diff
@@ -155,3 +153,85 @@ def search_container(self, cr_name, container_name):
             locators['resource.filter_containers'], container_name)
         strategy, value = locators['resource.select_container']
         return self.wait_until_element((strategy, value % container_name))
+
+    def list_vms(self, res_name):
+        """Lists vms of a particular compute resource.
+        Note: Currently lists only vms that show up on the first page.
```
Comment (elyezer) [inline robottelo/ui/computeresource.py]: Add a blank line between this line and the one above

File: robottelo.properties.sample (hunk 8 above)
Comment (kbidarkar) [inline robottelo.properties.sample]: ok, good idea, will do that and merge.

Comment (kbidarkar) [inline robottelo.properties.sample]: I mean merge the comment and field not the PR. :)

File: robottelo/ui/computeresource.py [hunk 10]
```diff
@@ -155,3 +153,87 @@ def search_container(self, cr_name, container_name):
...
+        self.click(tab_locators['tab_images'])
+        image_elements = self.find_elements(locators['resource.image_list'])
+        return [image.text for image in image_elements]
+
+    def vm_action_stop(self, res_name, vm_name, really):
+        """Stops a vm on the compute resource."""
+        self.click(self.search(res_name))
+        self.click(tab_locators['resource.tab_virtual_machines'])
+        strategy, value = locators['resource.vm_power_button']
+        button = self.find_element((strategy, value % (res_name, vm_name)))
```
Comment (rochacbruno) [inline robottelo/ui/computeresource.py]: This
```python
strategy, value = locators['resource.vm_power_button']
button = self.find_element((strategy, value % (res_name, vm_name)))
```

can now be rewritten as

```python
button = self.find_element(
    locators['resource.vm_power_button'] % (res_name, vm_name)
)
```

example:
```python
>>> loc = robottelo.ui.locators.Locator.XPATH('//div/%s/%s')
>>> print(x % ('foo', 'bar'))
<|strategy=xpath|value=//div/foo/bar|>

```

**Comment by rochacbruno** [comment]
ACK

(commented but not blocker, just informative)

PS: as coverage decreased, in future we can add tests for new set of config options in `tests/robottelo/test_config_settings.py`

File: robottelo/ui/computeresource.py [hunk 11]
```diff
@@ -71,9 +71,7 @@ def _configure_resource_provider(
                     'button'
                 ))
                 self.click(locators[button_locator])
-                self.find_element(
-                    locators[param_locator]
-                ).select_by_visible_text(parameter_value)
+                self.select(locators[param_locator], parameter_value)
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: We do not use `select` method directly anymore if we can, but `assign_value` instead

File: robottelo/ui/computeresource.py [hunk 12]
```diff
@@ -155,3 +153,87 @@ def search_container(self, cr_name, container_name):
...
         return self.wait_until_element((strategy, value % container_name))
+
+    def list_vms(self, res_name):
+        """Lists vms of a particular compute resource.
+
+        Note: Currently lists only vms that show up on the first page.
+        """
+        self.click(self.search(res_name))
+        self.click(tab_locators['resource.tab_virtual_machines'])
+        vm_elements = self.find_elements(locators['resource.vm_list'])
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: maybe `wait_for_element` will be more appropriate in that situation

File: robottelo/ui/computeresource.py [hunk 13]
```diff
@@ -155,3 +153,87 @@ def search_container(self, cr_name, container_name):
...
+        return [vm.text for vm in vm_elements]
+
+    def add_image(self, res_name, parameter_list):
+        """Adds an image to a compute resource."""
+        if parameter_list is None:
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: It doesn't seem that `parameter_list` can be None as it is required parameter, so that verification seems as redundant one

File: robottelo/ui/computeresource.py [hunk 14]
```diff
@@ -155,3 +153,87 @@ def search_container(self, cr_name, container_name):
...
+            return
+        self.click(self.search(res_name))
+        self.click(locators['resource.image_add'])
+        self.wait_until_element(locators['resource.image_name'])
+        for parameter_name, parameter_value, parameter_type in parameter_list:
+            param_locator = '.'.join((
+                'resource.image',
+                (parameter_name.lower()).replace(' ', '_')
+            ))
+            if parameter_type == 'field':
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: It doesn't seem that we need `parameter_type` logic anymore as we have `assign_value` method, so next statement should work for us:
`self.assign_value(locators[param_locator], parameter_value)`

File: robottelo/ui/computeresource.py [hunk 15]
```diff
@@ -155,3 +153,87 @@ def search_container(self, cr_name, container_name):
...
+        self.click(locators['resource.image_submit'])
+
+    def list_images(self, res_name):
+        """Lists images on Compute Resource.
+
+        Note: Currently lists only images that show up on the first page.
+        """
+        self.click(self.search(res_name))
+        self.click(tab_locators['tab_images'])
+        image_elements = self.find_elements(locators['resource.image_list'])
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: I would suggest `wait_until_element` here too

File: robottelo/ui/locators/base.py [hunk 16]
```diff
@@ -297,6 +298,42 @@
...
          "contains(@href, 'vms/') and contains(., '%s')]")),
+    "resource.get_by_name": (
+        By.XPATH,
+        "//a[not(contains(@href, 'search')) and contains(@href, '%s')]"),
+
+    # Locators under compute-resources vm tab.
+    "resource.vm_power_button": (
+        By.XPATH,
+        "//table[contains(@id, 'DataTables')]//a[contains(@data-id, '%s')"
+        "and .='%s']/../../td[6]//a"),
```
Comment (oshtaier) [inline robottelo/ui/locators/base.py]: Is any other way to avoid indexes here as it is very unstable way to find the elements?

File: robottelo/ui/locators/base.py [hunk 17]
```diff
@@ -297,6 +298,42 @@
...
+    "resource.vm_delete_button_dropdown": (
+        By.XPATH,
+        "//table[contains(@id, 'DataTables')]//a[contains(@data-id, '%s')"
+        "and .='%s']/../../td[6]//a[2]"),
```
Comment (oshtaier) [inline robottelo/ui/locators/base.py]: And here

File: robottelo/ui/computeresource.py (hunk 12 above)
Comment (kbidarkar) [inline robottelo/ui/computeresource.py]: I think we need new method self.find_elements will add one.

File: robottelo/ui/locators/base.py (hunk 16 above)
Comment (kbidarkar) [inline robottelo/ui/locators/base.py]: Ok, Even I try to avoid it, let me see what I come up with when I try one.

File: robottelo/ui/computeresource.py (hunk 12 above)
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: right, sorry, I missed `s` in the end

**Comment by oshtaier** [comment]
Please substitute `find_element` (without `s` in the end) on `wait_until_element`

**Comment by oshtaier** [comment]
I know that is hard to do, but any test result that prove implemented logic will be really helpful. At least that `find_elements` method works fine and return necessary data.

**Comment by oshtaier** [comment]
ACK pending last few comments

**Comment by kbidarkar** [comment]
Added a small testcase to test whether self.find_elements and other methods added to computeresource.py work fine.

```
def test_positive_retrieve_libvirt_vm_list(self):
        """Retrieve the Virtual machine list from Libvirt compute resource.
        @setup: Libvirt hostname and credentials.
        @steps:
        1. Select the created compute resource.
        2. Go to "Virtual Machines" tab.
        @Assert: The Virtual machine should be displayed.
        """
        vm_name = "bhqafvfgyzb.vhrth5g9zb"
        name = gen_string('alpha')
        with Session(self.browser) as session:
            make_resource(
                session,
                name=name,
                provider_type=FOREMAN_PROVIDERS['libvirt'],
                parameter_list=[
                    ['URL', self.current_libvirt_url, 'field']],
                orgs=[entities.Organization().create().name],
                org_select=True
            )
            search = self.compute_resource.search(name)
            self.assertIsNotNone(search)
            vms = self.compute_resource.list_vms(name)          
            self.assertTrue(vm_name in vms)
```
Currently just checking for a single vm to exist in the list fetched.
The  test result is as below:

```
 robottelo]$ nosetests -m test.positive_retrieve_libvirt_vm_list tests.foreman.ui.test_computeresource
.
----------------------------------------------------------------------
Ran 1 test in 86.626s

OK
```

**Comment by kbidarkar** [comment]
```
robottelo]$ nosetests -m test.positive_retrieve_libvirt_vm_list tests.foreman.ui.test_computeresource
.
----------------------------------------------------------------------
Ran 1 test in 92.505s

OK
```
Below is the code for above results.

```
        vm_name1 = "bhqafvfgyzb.vhrth5g9zb"
        vm_name2 = "cftvexapcfu.tsndwosyis"
        name = gen_string('alpha')
        with Session(self.browser) as session:
            make_resource(
                session,
                name=name,
                provider_type=FOREMAN_PROVIDERS['libvirt'],
                parameter_list=[
                    ['URL', self.current_libvirt_url, 'field']],
                orgs=[entities.Organization().create().name],
                org_select=True
            )
            search = self.compute_resource.search(name)
            self.assertIsNotNone(search)
            vms = self.compute_resource.list_vms(name)          
            self.assertTrue(vm_name1 in vms)
            self.assertTrue(vm_name2 in vms)
```

**Comment by oshtaier** [comment]
ACK
